src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -297,4 +301,60 @@ object PrecompiledContracts {
     }
   }
 
+  //Spec: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-197.md
+  // scalastyle: off
+  object Bn128Pairing extends PrecompiledContract {
+    private val wordLength = 32
+    private val inputLength = 6 * wordLength
+
+    val positiveResult = ByteUtils.padLeft(ByteString(1), wordLength)
+    val negativeResult = ByteString(Seq.fill(wordLength)(0.toByte).toArray)
+
+    def exec(inputData: ByteString): Option[ByteString] = {
+      if (inputData.length % inputLength != 0) {
+        None
+      } else {
+        getPairs(inputData.grouped(inputLength)).map{ pairs =>
+          if (PairingCheck.pairingCheck(pairs))
+            positiveResult
+          else
+            negativeResult
+        }
+      }
+    }
+
+    def gas(inputData: ByteString): BigInt = {
+      80000 * (inputData.length / inputLength) + 100000
+    }
+
+    // Method which stops reading another points if one of earlier ones failed (had invalid coordinates, or was not on
+    // BN128 curve
+    private def getPairs(bytes: Iterator[ByteString]): Option[Seq[G1G2Pair]] = {
+      var accum = List.empty[G1G2Pair]
+      while (bytes.hasNext) {
+        getPair(bytes.next) match {
+          case Some(part) => accum = part :: accum
+          case None => return None // scalastyle:ignore
```

Comment:
yep i need to agree  it is ugly :)it is not easy to create function to stop processing collection with first failed element fully functional way, it could probably be done with some lazy collections but i afraid intent would be lost there, and here intent is pretty clear